### PR TITLE
add flag to `is_connected` to allow user to see why provider connection failed

### LINKED ISSUE #2910: Show failure reason for `BaseProvider.is_connected()`
* Version: 6.0.0
* Python: 3.10
* OS: macky whacky
* `pip freeze`

```
ill provide if needed
```

### What was wrong?

This line right here officer: https://github.com/ethereum/web3.py/blob/master/web3/providers/base.py#L115
It failed on me, but it didn't tell me why.

**Also, note that `assert` statements in production python code get swept-out when global `__debug__` is set to `False`.**

This whole `assert` line should be a raised exception imo.

### How can it be fixed?

Raise an exception with `error` in the message.

---
**Note:** We prefer to use issues to track our work. If you think you've encountered a bug in web3py or 
have a feature request, you're in the right place. If you have implementation or usage questions, 
please refer to our [documentation](https://web3py.readthedocs.io/en/latest/) and/or join the conversation 
on [discord](https://discord.gg/GHryRvPB84).

### PR DESCRIPTION
**Pull request by pacrob**
### What was wrong?

`w3.is_connected()` just returns True or False. It would be helpful to know why the connection failed if it is False.
Closes #2910 

### How was it fixed?

Added a `raise_if_false` flag to the `is_connected` method and a new `ProviderConnectionError` which inherits from `Web3Exception`.

While adding tests, did a little cleanup on other provider testing.

### Todo:
- [x] Add entry to the [release notes](https://github.com/ethereum/web3.py/blob/master/newsfragments/README.md)

#### Cute Animal Picture

image

### CODE REVIEW

**Review by kclowes — APPROVED** [review]
Looks good! I had a naming nit, and one question about the cache size test, but feel free to take or leave!

File: docs/internals.rst [hunk 1]
```diff
@@ -101,13 +101,17 @@ setting the middlewares the provider should use.
       the JSON-RPC method being called.
 
 
-.. py:method:: BaseProvider.is_connected()
+.. py:method:: BaseProvider.is_connected(raise_if_false=False)
```
[the remark this answers is not in the document]
Comment (pacrob) [reply docs/internals.rst]: I like `show_traceback`. Changed.

File: docs/internals.rst [hunk 2]
```diff
@@ -101,13 +101,17 @@ setting the middlewares the provider should use.
...
 
     This function should return ``True`` or ``False`` depending on whether the
     provider should be considered *connected*.  For example, an IPC socket
     based provider should return ``True`` if the socket is open and ``False``
     if the socket is closed.
 
+    If set to ``True``, the optional ``raise_if_false`` boolean should raise a
```
[the remark this answers is not in the document]
Comment (pacrob) [reply docs/internals.rst]: fixed

File: tests/core/providers/test_async_http_provider.py [hunk 3]
```diff
@@ -83,5 +90,5 @@ async def test_user_provided_session() -> None:
     session = ClientSession()
     provider = AsyncHTTPProvider(endpoint_uri=URI)
     cached_session = await provider.cache_async_session(session)
-    assert len(request._async_session_cache) == 1
+    assert len(request._async_session_cache) == 2
```
[the remark this answers is not in the document]
Comment (pacrob) [reply tests/core/providers/test_async_http_provider.py]: The `is_connected`s I added were added to the cache. Updated to clear the cache at the end of the test.

**Pull request merged**